types/lib/linear.go

```diff
@@ -0,0 +1,306 @@
+
+	if opt.LengthKey == nil {
+		for ; ; i++ {
+			if bytes.Equal(opt.ElemKey, []byte{i}) || bytes.Equal(opt.TopKey, []byte{i}) {
+				continue
+			}
+			opt.LengthKey = []byte{i}
+			break
+		}
+	}
+
+	if opt.ElemKey == nil {
+		for ; ; i++ {
+			if bytes.Equal(opt.LengthKey, []byte{i}) || bytes.Equal(opt.TopKey, []byte{i}) {
+				continue
+			}
+			opt.ElemKey = []byte{i}
+			break
+		}
+	}
+
+	if opt.TopKey == nil {
+		for ; ; i++ {
+			if bytes.Equal(opt.LengthKey, []byte{i}) || bytes.Equal(opt.ElemKey, []byte{i}) {
+				continue
+			}
+			opt.TopKey = []byte{i}
+			break
+		}
+	}
+
+	opt.initialized = true
+}
+
+// NewLinear constructs new Linear
+func NewLinear(cdc *wire.Codec, store sdk.KVStore, opts ...*LinearOptions) Linear {
+	var opt *LinearOptions
+	switch len(opts) {
+	case 0:
+		opt = cachedDefaultLinearOptions
+	case 1:
+		opt = opts[0]
+		if !opt.initialized {
+			opt.init()
+		}
+	default:
+		panic("improper use of NewList, can only have one LinearOptions")
+	}
+
+	return Linear{
+		cdc:   cdc,
+		store: store,
+		opt:   opt,
+	}
+}
+
+// List is a Linear interface that provides list-like functions
+// It panics when the element type cannot be (un/)marshalled by the codec
+type List interface {
+
+	// Len() returns the length of the list
+	// The length is only increased by Push() and not decreased
+	// List dosen't check if an index is in bounds
+	// The user should check Len() before doing any actions
+	Len() uint64
+
+	// Get() returns the element by its index
+	Get(uint64, interface{}) error
+
+	// Set() stores the element to the given position
+	// Setting element out of range will break length counting
+	// Use Push() instead of Set() to append a new element
+	Set(uint64, interface{})
+
+	// Delete() deletes the element in the given position
+	// Other elements' indices are preserved after deletion
+	// Panics when the index is out of range
+	Delete(uint64)
+
+	// Push() inserts the element to the end of the list
+	// It will increase the length when it is called
+	Push(interface{})
+
+	// Iterate*() is used to iterate over all existing elements in the list
+	// Return true in the continuation to break
+	// The second element of the continuation will indicate the position of the element
+	// Using it with Get() will return the same one with the provided element
+
+	// CONTRACT: No writes may happen within a domain while iterating over it.
+	Iterate(interface{}, func(uint64) bool)
+}
+
+// NewList constructs new List
+func NewList(cdc *wire.Codec, store sdk.KVStore, opts ...*LinearOptions) List {
+	return NewLinear(cdc, store, opts...)
+}
+
+// Key for the length of the list
+func (m Linear) LengthKey() []byte {
+	return m.opt.LengthKey
+}
+
+// Key for the elements of the list
+func (m Linear) ElemKey(index uint64) []byte {
+	return append(m.opt.ElemKey, []byte(fmt.Sprintf("%020d", index))...)
+}
+
+// Len implements List
+func (m Linear) Len() (res uint64) {
+	bz := m.store.Get(m.LengthKey())
+	if bz == nil {
```

Comment:
Can we initially set the length key to zero so we don't need to check this? I think it's a bit more obvious if we ensure all store invariants in the `NewXYZ()` functions.

Not a big deal though.

Reply:
A Linear is constructed every time when it is needed, so It will be hard to know is this the first call(so it need to set the value to 0) or not(skip).